File: elo/elo.py
from models import Tournament, TournamentQuestion

_retries = 0


class Elo:
    def __init__(self, tournament: Tournament):
        self.tournament_id = tournament.id
        self.initial_k = tournament.initial_k
        self.minimum_k = tournament.minimum_k
        self.std_dev_multiplier = tournament.std_dev_multiplier
        self.initial_phase_matches = tournament.initial_phase_matches
        self.transition_phase_matches = tournament.transition_phase_matches
        self.top_n = tournament.top_n
        self.band_size = tournament.band_size

    def _select_two_unqualified(self) -> tuple[int, int]:
        first = TournamentQuestion.get_random_question(self.tournament_id, max_matches=self.initial_phase_matches - 1)
        second = TournamentQuestion.get_random_question(self.tournament_id, max_matches=self.initial_phase_matches - 1)
        return first.question_id, second.question_id

    def _select_two_qualified(self) -> tuple[int, int]:
        threshold = self.calculate_threshold()
        first = TournamentQuestion.get_random_question(tournament_id=self.tournament_id, min_rating=threshold)
        second = TournamentQuestion.get_random_question(tournament_id=self.tournament_id, min_rating=threshold)
        return first.question_id, second.question_id

    def select_pair(self) -> tuple[int, int]:
        global _retries
        unqualified_count = TournamentQuestion.count_unqualified_questions(
            self.tournament_id, self.initial_phase_matches
        )
        if unqualified_count > 1:
            first, second = self._select_two_unqualified()
        elif unqualified_count == 1:
            first = TournamentQuestion.get_random_question(
                self.tournament_id, max_matches=self.initial_phase_matches - 1
            )
            second = TournamentQuestion.get_random_question(self.tournament_id)
        else:
            first, second = self._select_two_qualified()

        if first == second:
            _retries += 1
            return self.select_pair()

        return first, second

    def _calculate_k_factor(self, item: TournamentQuestion) -> float:
        if item.matches < self.initial_phase_matches:
            return self.initial_k
        elif item.matches < self.transition_phase_matches:
            return self.initial_k / 2
        else:
            return self.minimum_k

    def record_winner(self, winner_id: int, loser_id: int) -> None:
        winner = TournamentQuestion.find(self.tournament_id, winner_id)
        loser = TournamentQuestion.find(self.tournament_id, loser_id)

        winner.matches += 1
        loser.matches += 1
        winner.wins += 1

        expected_winner = 1 / (1 + 10 ** ((loser.rating - winner.rating) / 400))
        winner_k = self._calculate_k_factor(winner)
        loser_k = self._calculate_k_factor(loser)
        k_factor = (winner_k + loser_k) / 2

        rating_change = k_factor * (1 - expected_winner)

        winner.rating += rating_change
        loser.rating -= rating_change

        winner.save()
        loser.save()

    def get_top_items(self, n: int = 100) -> list:
        return TournamentQuestion.get_top_questions(self.tournament_id, n)

    def calculate_threshold(self) -> float:
        ratings_count, std_dev = TournamentQuestion.get_stats_for_qualified(
            self.tournament_id, self.initial_phase_matches
        )
        if ratings_count < self.top_n:
            return float("-inf")

        top_n_threshold = TournamentQuestion.get_rating_at_position(self.tournament_id, self.top_n)
        return top_n_threshold - (self.std_dev_multiplier * std_dev)

    def get_questions_stats(self, q1_id: int, q2_id: int) -> list[dict]:
        q1 = TournamentQuestion.find(self.tournament_id, q1_id)
        q2 = TournamentQuestion.find(self.tournament_id, q2_id)
        return [{"matches": q1.matches, "wins": q1.wins}, {"matches": q2.matches, "wins": q2.wins}]

    def get_statistics(self) -> dict:
        global _retries
        threshold = self.calculate_threshold()
        above_threshold_count = TournamentQuestion.count_questions_above_threshold(self.tournament_id, threshold)
        rating_distribution = TournamentQuestion.get_rating_distribution(self.tournament_id)
        unqualified_count = TournamentQuestion.count_unqualified_questions(
            self.tournament_id, self.initial_phase_matches
        )
        match_counts = TournamentQuestion.get_match_counts(self.tournament_id)

        return {
            "current_threshold": threshold,
            "above_threshold": above_threshold_count,
            "unqualified": unqualified_count,
            "distribution": rating_distribution,
            "retries": _retries,
            "total_matches": match_counts["matches"],
            "total_wins": match_counts["wins"],
        }

File: internal/elo/elo.go
package elo

import (
	"fmt"
	"math"
	"questions-vote/internal/models"
	"sync"
)

var (
	retries int
	mu      sync.Mutex
)

// ELO implements the ELO rating system for tournament questions
type ELO struct {
	TournamentID           int
	InitialK               float64
	MinimumK               float64
	StdDevMultiplier       float64
	InitialPhaseMatches    int
	TransitionPhaseMatches int
	TopN                   int
	BandSize               int
	
	tournamentQuestionRepo *models.TournamentQuestionRepository
}

// New creates a new ELO instance from a tournament
func New(tournament *models.Tournament) *ELO {
	return &ELO{
		TournamentID:           tournament.ID,
		InitialK:               tournament.InitialK,
		MinimumK:               tournament.MinimumK,
		StdDevMultiplier:       tournament.StdDevMultiplier,
		InitialPhaseMatches:    tournament.InitialPhaseMatches,
		TransitionPhaseMatches: tournament.TransitionPhaseMatches,
		TopN:                   tournament.TopN,
		BandSize:               tournament.BandSize,
		tournamentQuestionRepo: models.NewTournamentQuestionRepository(),
	}
}

// SelectPair selects two questions for comparison based on ELO algorithm
func (e *ELO) SelectPair() (int, int, error) {
	mu.Lock()
	defer mu.Unlock()
	
	return e.selectPairInternal()
}

func (e *ELO) selectPairInternal() (int, int, error) {
	unqualifiedCount, err := e.tournamentQuestionRepo.CountUnqualifiedQuestions(e.TournamentID, e.InitialPhaseMatches)
	if err != nil {
		return 0, 0, fmt.Errorf("failed to count unqualified questions: %w", err)
	}
	
	var first, second int
	
	if unqualifiedCount > 1 {
		first, second, err = e.selectTwoUnqualified()
	} else if unqualifiedCount == 1 {
		first, err = e.selectUnqualified()
		if err != nil {
			return 0, 0, err
		}
		second, err = e.selectAny()
	} else {
		first, second, err = e.selectTwoQualified()
	}
	
	if err != nil {
		return 0, 0, err
	}
	
	if first == second {
		retries++
		return e.selectPairInternal()
	}
	
	return first, second, nil
}

// selectTwoUnqualified selects two unqualified questions
func (e *ELO) selectTwoUnqualified() (int, int, error) {
	maxMatches := e.InitialPhaseMatches - 1
	
	first, err := e.tournamentQuestionRepo.GetRandomQuestion(e.TournamentID, 0, math.MaxFloat64, maxMatches)
	if err != nil {
		return 0, 0, fmt.Errorf("failed to get first unqualified question: %w", err)
	}
	
	second, err := e.tournamentQuestionRepo.GetRandomQuestion(e.TournamentID, 0, math.MaxFloat64, maxMatches)
	if err != nil {
		return 0, 0, fmt.Errorf("failed to get second unqualified question: %w", err)
	}
	
	return first.QuestionID, second.QuestionID, nil
}

// selectUnqualified selects one unqualified question
func (e *ELO) selectUnqualified() (int, error) {
	maxMatches := e.InitialPhaseMatches - 1
	
	question, err := e.tournamentQuestionRepo.GetRandomQuestion(e.TournamentID, 0, math.MaxFloat64, maxMatches)
	if err != nil {
		return 0, fmt.Errorf("failed to get unqualified question: %w", err)
	}
	
	return question.QuestionID, nil
}

// selectAny selects any question
func (e *ELO) selectAny() (int, error) {
	question, err := e.tournamentQuestionRepo.GetRandomQuestion(e.TournamentID, 0, math.MaxFloat64, math.MaxInt32)
	if err != nil {
		return 0, fmt.Errorf("failed to get any question: %w", err)
	}
	
	return question.QuestionID, nil
}

// selectTwoQualified selects two qualified questions above threshold
func (e *ELO) selectTwoQualified() (int, int, error) {
	threshold, err := e.CalculateThreshold()
	if err != nil {
		return 0, 0, fmt.Errorf("failed to calculate threshold: %w", err)
	}
	
	first, err := e.tournamentQuestionRepo.GetRandomQuestion(e.TournamentID, threshold, math.MaxFloat64, math.MaxInt32)
	if err != nil {
		return 0, 0, fmt.Errorf("failed to get first qualified question: %w", err)
	}
	
	second, err := e.tournamentQuestionRepo.GetRandomQuestion(e.TournamentID, threshold, math.MaxFloat64, math.MaxInt32)
	if err != nil {
		return 0, 0, fmt.Errorf("failed to get second qualified question: %w", err)
	}
	
	return first.QuestionID, second.QuestionID, nil
}

// calculateKFactor calculates the K-factor for a tournament question
func (e *ELO) calculateKFactor(tq *models.TournamentQuestion) float64 {
	if tq.Matches < e.InitialPhaseMatches {
		return e.InitialK
	} else if tq.Matches < e.TransitionPhaseMatches {
		return e.InitialK / 2
	} else {
		return e.MinimumK
	}
}

// RecordWinner records the winner and updates ELO ratings
func (e *ELO) RecordWinner(winnerID, loserID int) error {
	winner, err := e.tournamentQuestionRepo.Find(e.TournamentID, winnerID)
	if err != nil {
		return fmt.Errorf("failed to find winner: %w", err)
	}
	
	loser, err := e.tournamentQuestionRepo.Find(e.TournamentID, loserID)
	if err != nil {
		return fmt.Errorf("failed to find loser: %w", err)
	}
	
	// Update match counts
	winner.Matches++
	loser.Matches++
	winner.Wins++
	
	// Calculate expected winner probability
	expectedWinner := 1.0 / (1.0 + math.Pow(10, (loser.Rating-winner.Rating)/400))
	
	// Calculate K-factors
	winnerK := e.calculateKFactor(winner)
	loserK := e.calculateKFactor(loser)
	kFactor := (winnerK + loserK) / 2
	
	// Calculate rating change
	ratingChange := kFactor * (1 - expectedWinner)
	
	// Update ratings
	winner.Rating += ratingChange
	loser.Rating -= ratingChange
	
	// Save changes
	err = e.tournamentQuestionRepo.Save(winner)
	if err != nil {
		return fmt.Errorf("failed to save winner: %w", err)
	}
	
	err = e.tournamentQuestionRepo.Save(loser)
	if err != nil {
		return fmt.Errorf("failed to save loser: %w", err)
	}
	
	return nil
}

// GetTopItems returns the top N questions by rating
func (e *ELO) GetTopItems(n int) ([]*models.TournamentQuestion, error) {
	return e.tournamentQuestionRepo.GetTopQuestions(e.TournamentID, n)
}

// CalculateThreshold calculates the rating threshold for question selection
func (e *ELO) CalculateThreshold() (float64, error) {
	ratingsCount, stdDev, err := e.tournamentQuestionRepo.GetStatsForQualified(e.TournamentID, e.InitialPhaseMatches)
	if err != nil {
		return 0, fmt.Errorf("failed to get stats for qualified: %w", err)
	}
	
	if ratingsCount < e.TopN {
		return math.Inf(-1), nil
	}
	
	topNThreshold, err := e.tournamentQuestionRepo.GetRatingAtPosition(e.TournamentID, e.TopN)
	if err != nil {
		return 0, fmt.Errorf("failed to get rating at position: %w", err)
	}
	
	return topNThreshold - (e.StdDevMultiplier * stdDev), nil
}

// GetQuestionsStats returns statistics for two questions
func (e *ELO) GetQuestionsStats(q1ID, q2ID int) ([]models.QuestionStats, error) {
	q1, err := e.tournamentQuestionRepo.Find(e.TournamentID, q1ID)
	if err != nil {
		return nil, fmt.Errorf("failed to find question 1: %w", err)
	}
	
	q2, err := e.tournamentQuestionRepo.Find(e.TournamentID, q2ID)
	if err != nil {
		return nil, fmt.Errorf("failed to find question 2: %w", err)
	}
	
	return []models.QuestionStats{
		{Wins: q1.Wins, Matches: q1.Matches},
		{Wins: q2.Wins, Matches: q2.Matches},
	}, nil
}

// GetStatistics returns comprehensive tournament statistics
func (e *ELO) GetStatistics() (map[string]interface{}, error) {
	mu.Lock()
	currentRetries := retries
	mu.Unlock()
	
	threshold, err := e.CalculateThreshold()
	if err != nil {
		return nil, fmt.Errorf("failed to calculate threshold: %w", err)
	}
	
	aboveThresholdCount, err := e.tournamentQuestionRepo.CountQuestionsAboveThreshold(e.TournamentID, threshold)
	if err != nil {
		return nil, fmt.Errorf("failed to count questions above threshold: %w", err)
	}
	
	ratingDistribution, err := e.tournamentQuestionRepo.GetRatingDistribution(e.TournamentID, 20)
	if err != nil {
		return nil, fmt.Errorf("failed to get rating distribution: %w", err)
	}
	
	unqualifiedCount, err := e.tournamentQuestionRepo.CountUnqualifiedQuestions(e.TournamentID, e.InitialPhaseMatches)
	if err != nil {
		return nil, fmt.Errorf("failed to count unqualified questions: %w", err)
	}
	
	totalMatches, totalWins, err := e.tournamentQuestionRepo.GetMatchCounts(e.TournamentID)
	if err != nil {
		return nil, fmt.Errorf("failed to get match counts: %w", err)
	}
	
	return map[string]interface{}{
		"current_threshold":  threshold,
		"above_threshold":    aboveThresholdCount,
		"unqualified":        unqualifiedCount,
		"distribution":       ratingDistribution,
		"retries":            currentRetries,
		"total_matches":      totalMatches,
		"total_wins":         totalWins,
	}, nil
}

// GetRetryCount returns the current retry count (for testing)
func GetRetryCount() int {
	mu.Lock()
	defer mu.Unlock()
	return retries
}

// ResetRetryCount resets the retry count (for testing)
func ResetRetryCount() {
	mu.Lock()
	defer mu.Unlock()
	retries = 0
}